modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/logical/LogicalOrToUnionRule.java

```diff
@@ -155,6 +167,59 @@ private RelNode createUnionAll(RelOptCluster cluster, RelNode input, RexNode op1
             .build();
     }
 
+    /** */
+    private static boolean preMatch(IgniteLogicalTableScan scan) {
+        RexNode cond = scan.condition();
+
+        if (cond == null ||
```

Comment:
I have some proposals:
1. Let's keep only checks from this condition in `preMatch`.
2. Move index collation check to the `onMatch` method.
3. Check index collation for both operands and proceed only if both operands match any index.
4. Since we can effectively use index only if filtering first field of index key - make a bitset of first fields of all indexes and proceed only if the intersection of fields in condition and first fields of indexes is not empty.

Reply:
done